File: app/src/main/java/ru/nikshlykov/englishwordsapp/ui/viewmodels/SubgroupDataViewModel.java
package ru.nikshlykov.englishwordsapp.ui.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import ru.nikshlykov.englishwordsapp.db.GroupsRepository;
import ru.nikshlykov.englishwordsapp.db.subgroup.Subgroup;

public class SubgroupDataViewModel extends ViewModel
        implements GroupsRepository.OnSubgroupInsertedListener, GroupsRepository.OnSubgroupUpdatedListener {

    private GroupsRepository groupsRepository;

    private LiveData<Subgroup> subgroup;

    private MutableLiveData<Boolean> subgroupIsInsertedOrUpdated;

    public SubgroupDataViewModel(GroupsRepository groupsRepository) {
        this.groupsRepository = groupsRepository;
        subgroup = new MutableLiveData<>();
        subgroupIsInsertedOrUpdated = new MutableLiveData<>();
    }

    public void setSubgroup(long subgroupId) {
        subgroup = groupsRepository.getLiveDataSubgroupById(subgroupId);
    }

    public LiveData<Subgroup> getSubgroup() {
        return subgroup;
    }

    public LiveData<Boolean> getSubgroupIsInsertedOrUpdated() {
        return subgroupIsInsertedOrUpdated;
    }

    public void insertOrUpdateSubgroup(String subgroupName) {
        Subgroup subgroupToUpdate = this.subgroup.getValue();
        if (subgroupToUpdate != null) {
            subgroupToUpdate.name = subgroupName;
            groupsRepository.update(subgroupToUpdate, this);
        } else {
            groupsRepository.insertSubgroup(subgroupName, this);
        }
    }

    @Override
    public void onSubgroupInserted(long subgroupId) {
        if (subgroupId != 0L) {
            subgroupIsInsertedOrUpdated.postValue(true);
        }
    }

    @Override
    public void onSubgroupUpdated(boolean isSubgroupUpdated) {
        if (isSubgroupUpdated) {
            subgroupIsInsertedOrUpdated.postValue(true);
        }
    }
}

File: app/src/main/java/ru/nikshlykov/englishwordsapp/ui/viewmodels/SubgroupDataViewModel.kt
package ru.nikshlykov.englishwordsapp.ui.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import ru.nikshlykov.englishwordsapp.db.subgroup.Subgroup
import ru.nikshlykov.englishwordsapp.domain.interactors.AddSubgroupInteractor
import ru.nikshlykov.englishwordsapp.domain.interactors.GetSubgroupInteractor
import ru.nikshlykov.englishwordsapp.domain.interactors.UpdateSubgroupInteractor

class SubgroupDataViewModel(
  private val getSubgroupInteractor: GetSubgroupInteractor,
  private val addSubgroupInteractor: AddSubgroupInteractor,
  private val updateSubgroupInteractor: UpdateSubgroupInteractor
) : ViewModel() {
  private val _subgroup: MutableLiveData<Subgroup> = MutableLiveData()

  val subgroup: LiveData<Subgroup> = _subgroup

  val subgroupIsInsertedOrUpdated: MutableLiveData<Boolean> = MutableLiveData()


  fun loadSubgroup(subgroupId: Long) {
    viewModelScope.launch {
      _subgroup.value = getSubgroupInteractor.getSubgroupById(subgroupId)
    }
  }

  fun getSubgroupIsInsertedOrUpdated(): LiveData<Boolean> {
    return subgroupIsInsertedOrUpdated
  }

  fun addOrUpdateSubgroup(subgroupName: String?) {
    val subgroupToUpdate = subgroup.value
    if (subgroupToUpdate != null) {
      if (subgroupName != null) {
        subgroupToUpdate.name = subgroupName
        viewModelScope.launch {
          val subgroupUpdated = updateSubgroupInteractor.updateSubgroup(subgroupToUpdate)
          if (subgroupUpdated == 1) {
            subgroupIsInsertedOrUpdated.postValue(true)
          }
        }
      } else {
        // TODO тоже самое, что и ниже
      }
    } else {
      if (subgroupName != null) {
        viewModelScope.launch {
          val newSubgroupId = addSubgroupInteractor.addSubgroup(subgroupName)
          if (newSubgroupId != 0L) {
            subgroupIsInsertedOrUpdated.postValue(true)
          }
        }
      } else {
        // TODO сделать Toast для сообщения о пустом поле. Или DataBinding на поле.
        //  Или заставить это обрабатывать Вьюху.
      }
    }
  }
}